lib/http2.c

```diff
@@ -1999,26 +2000,33 @@ static ssize_t cf_h2_send(struct Curl_cfilter *cf, struct Curl_easy *data,
     rv = nghttp2_session_resume_data(ctx->h2, stream->stream_id);
     if(nghttp2_is_fatal(rv)) {
       *err = CURLE_SEND_ERROR;
-      len = -1;
+      nwritten = -1;
       goto out;
     }
     result = h2_session_send(cf, data);
     if(result) {
       *err = result;
-      len = -1;
+      nwritten = -1;
       goto out;
     }
-    len -= stream->upload_len;
 
-    /* Nullify here because we call nghttp2_session_send() and they
-       might refer to the old buffer. */
+    /* The callback for stream input updates `upload_len` with what it
+     * has provided to nghttp2. If it has *not* been called, `upload_len`
+     * is unchanged and we should report a block on send. */
+    if(stream->upload_len == len) {
+      *err = CURLE_AGAIN;
+      nwritten = -1;
+      goto out;
+    }
```

Comment:
I am not sure about this logic, I notice there's a check for !len that comes later in the `ifdef DEBUG_HTTP2` section. Also the two len references in that section should be updated to nwritten

Reply:
Good catch. The `len` use below should use `nwritten`.

Reply:
Updated in recent push.

Reply:
Thanks but I meant the new logic to return eagain rather than continue on and call resume, should_close_session etc and !nwritten eval (which will never be true now)